[ACM-4613]: Adding instructions to enable ARM control planes for HyperShift on AWS (#4865)
open-cluster-management/rhacm-docs: clusters/hosted_control_planes/managing_hosted_aws.adoc

- `kubeconfig` secret: `<hostingNamespace>-<name>-admin-kubeconfig` (clusters-hypershift-demo-admin-kubeconfig)
- `kubeadmin` password secret: `<hostingNamespace>-<name>-kubeadmin-password` (clusters-hypershift-demo-kubeadmin-password)

[#hosted-cluster-arm-aws]
== Enabling a hosted cluster on {ocp-short} ARM

> You can enable an x86 control plane to operate with an {ocp} ARM data plane in a hosted control planes environment. This feature is available for hosted control planes on AWS only.

Reviewer: ```suggestion
You can enable an ARM hosted control plane to operate within an {ocp} ARM data plane in a management cluster environment. This feature is available for hosted control planes on AWS only.
```

Author: Thanks for the clarifications!

.Prerequisites

Before you begin, you must meet the following prerequisites:

* You must have an {ocp-short} ARM management cluster. 

> // Will most users know where to get an OCP ARM management cluster? If not, we should point them in the right direction.

Reviewer: Could we link up to ARM docs here? I'm sure Kelly might know if we have docs that help with this. If not, I can check as well.

Author: Sure! I'll ask Kelly which link to use.

Reviewer 2: There is no specific "Install on OCP on ARM" doc becuase its just the normal installation with a couple changes. Mostly just picking the correct payload in the hybrid cloud console https://console.redhat.com/openshift/install/aws/arm

* You must have a HyperShift Operator that is built on ARM. You can obtain a HyperShift Operator in one of two ways:

> //I didn't include the option to download the Operator from https://quay.io/repository/acm-d/hypershift-rhel8-operator because it looks like that option requires onboarding, which doesn't seem appropriate for users.

Reviewer: Agree. I believe the ACM route is how SD is installing HyperShift in ROSA; so to me, that is more of the managed services side for HyperShift. 

These docs seem more of the flavor of self-managed.

** Go to the link:https://quay.io/repository/hypershift/hypershift-operator[hypershift/hypershift-operator repository] and select the build that has the `4.13-arm64` tag. 
** Build your own HyperShift Operator by modifying the HyperShift Dockerfile to build from ARM images.

> //Do we want to mention this option in the product docs? If so, we should likely explain how to build the Operator by modifying the Dockerfile.

Reviewer: Hmm maybe not. I'm inquiring on the ARM side. I'm not sure the modifications I have made in the past are open to users outside of Red Hat.

Reviewer: I was told we probably should not include that. I believe the registries we were using internally to change the Dockerfile are CI repos.

Author: Thanks for confirming! I'll remove that.

.Procedure

To run a hosted cluster on {ocp-short} ARM, take the following steps:

> . Install the ARM HyperShift Operator on the management cluster.

Reviewer: ```suggestion
. Install the ARM HyperShift Operator on the management cluster overriding the default HyperShift Operator image.
```

> // Is this step straightforward for most people, or do we need to tell users how to install the Operator on the management cluster?

Reviewer: For example, through the HyperShift CLI:
```
hypershift install \
--oidc-storage-provider-s3-bucket-name $BUCKET_NAME \
--oidc-storage-provider-s3-credentials $AWS_CREDS \
--oidc-storage-provider-s3-region $REGION \
--hypershift-image quay.io/hypershift/hypershift-operator:4.13-arm64
```

Author: @bryan-cox Thanks for the example! Is it safe to assume that `$BUCKET_NAME`, `$AWS_CREDS`, and `$REGION` are variables that the user will need to replace?

Reviewer: Yeah any of those with a $ would need to be replaced by the user.

. Create a hosted cluster that overrides the default release image with a multi-architecture release image.

> // Can we add more details here?

Reviewer: For example, through the HyperShift CLI (this will add a default NodePool through the `--node-pool-replicas` flag):
```
hypershift create cluster aws \ 
--name $CLUSTER_NAME \
--node-pool-replicas=$NODEPOOL_REPLICAS \
--base-domain $BASE_DOMAIN \
--pull-secret $PULL_SECRET \
--aws-creds $AWS_CREDS \
--region $REGION \
--release-image quay.io/openshift-release-dev/ocp-release:4.13.0-rc.0-multi

```

Author: Thanks! Same question about variable text. Is it safe to assume that anything starting with a dollar sign is a variable the the user needs to replace?

Reviewer: Yeah, the user needs to replace.

. Add an x86 `NodePool` object to the hosted cluster.

> // Can we add more details here?

Reviewer: For example, through the HyperShift CLI:
```
hypershift create nodepool aws \
--cluster-name $CLUSTER_NAME \
--name $NODEPOOL_NAME \
--node-count=$NODEPOOL_REPLICAS
```

- `kubeconfig` secret: `<hostingNamespace>-<name>-admin-kubeconfig` (clusters-hypershift-demo-admin-kubeconfig)
- `kubeadmin` password secret: `<hostingNamespace>-<name>-kubeadmin-password` (clusters-hypershift-demo-kubeadmin-password)

[#hosted-cluster-arm-aws]
== Enabling hosted control planes on an ARM64 {ocp-short} cluster

> You can enable an ARM64-hosted control plane to operate with an {ocp} ARM64 data plane in a management cluster environment. This feature is available for hosted control planes on AWS only.

Reviewer 3: Can we use {ocp-short} here as well?

> .Prerequisites

Reviewer 3: ```suggestion
[#prerequisites-hosted-arm]
== Prerequisites 
```

Just another thing we seem to do differently compared to OCP docs

Before you begin, you must meet the following prerequisites:

* You must have an {ocp-short} cluster that was installed on a 64-bit ARM infrastructure. For more information, see link:https://console.redhat.com/openshift/install/aws/arm[Create an OpenShift Cluster: AWS (ARM)].
* You must have a HyperShift Operator that is built on a 64-bit ARM infrastructure. You can obtain a HyperShift Operator by going to the link:https://quay.io/repository/hypershift/hypershift-operator[hypershift/hypershift-operator repository] and selecting the build that has the `4.13-arm64` tag. 

> .Procedure

Reviewer 3: I know this is how to indicate a procedure in OCP docs. We don't use that so you can leave out this line.

- `kubeconfig` secret: `<hostingNamespace>-<name>-admin-kubeconfig` (clusters-hypershift-demo-admin-kubeconfig)
- `kubeadmin` password secret: `<hostingNamespace>-<name>-kubeadmin-password` (clusters-hypershift-demo-kubeadmin-password)

[#hosted-cluster-arm-aws]

> == Enabling a hosted cluster on {ocp-short} ARM

Reviewer 2: ```suggestion
== Enabling a hosted cluster on a ARM64 {ocp-short} cluster
```

> You can enable an ARM-hosted control plane to operate with an {ocp} ARM data plane in a management cluster environment. This feature is available for hosted control planes on AWS only.

Reviewer 2: ```suggestion
You can enable an ARM64-hosted control plane to operate with an {ocp} ARM64 data plane in a management cluster environment. This feature is available for hosted control planes on AWS only.
```

.Prerequisites

Before you begin, you must meet the following prerequisites:

> * You must have an {ocp-short} ARM management cluster. 

Reviewer 2: ```suggestion
* You must have an ARM64 based {ocp-short} cluster. 
```
If you want to be more specific you can say something like 
"You have a {ocp-short} cluster that was installed on a 64-bit ARM infrastructure"

// Will most users know where to get an OCP ARM management cluster? If not, we should point them in the right direction.

> * You must have a HyperShift Operator that is built on ARM. You can obtain a HyperShift Operator by going to the link:https://quay.io/repository/hypershift/hypershift-operator[hypershift/hypershift-operator repository] and selecting the build that has the `4.13-arm64` tag. 

Reviewer 2: ```suggestion
* You must have a HyperShift Operator that is built on a 64-bit ARM infrastructure. You can obtain a HyperShift Operator by going to the link:https://quay.io/repository/hypershift/hypershift-operator[hypershift/hypershift-operator repository] and selecting the build that has the `4.13-arm64` tag. 
```

.Procedure

> To run a hosted cluster on {ocp-short} ARM, take the following steps:

Reviewer 2: ```suggestion
To run a hosted cluster on a ARM64 {ocp-short} cluster, take the following steps:
```

> . Install the ARM HyperShift Operator on the management cluster to override the default HyperShift Operator image.

Reviewer 2: ```suggestion
. Install the HyperShift Operator for ARM64 on the management cluster to override the default HyperShift Operator image.
```

+
For example, through the hosted control planes (`hypershift`) CLI, enter the following commands, being careful to replace the bucket name, AWS credentials, and region with your information:
+
----
hypershift install \
--oidc-storage-provider-s3-bucket-name $BUCKET_NAME \
--oidc-storage-provider-s3-credentials $AWS_CREDS \
--oidc-storage-provider-s3-region $REGION \
--hypershift-image quay.io/hypershift/hypershift-operator:4.13-arm64
----

. Create a hosted cluster that overrides the default release image with a multi-architecture release image.
+
For example, through the hosted control planes (`hypershift`) CLI, enter the following commands, being careful to replace the cluster name, node pool replicas, base domain, pull secret, AWS credentials, and region with your information:
+
----
hypershift create cluster aws \ 
--name $CLUSTER_NAME \
--node-pool-replicas=$NODEPOOL_REPLICAS \
--base-domain $BASE_DOMAIN \
--pull-secret $PULL_SECRET \
--aws-creds $AWS_CREDS \
--region $REGION \
--release-image quay.io/openshift-release-dev/ocp-release:4.13.0-rc.0-multi
----
+
This example adds a default `NodePool` object through the `--node-pool-replicas` flag.

> . Add an x86 `NodePool` object to the hosted cluster.

Reviewer 2: ```suggestion
. Add an x86_64 `NodePool` object to the hosted cluster.
```
Def check with Bryan on this one, bc the value could be `x86` on its own, but I rarly see it just as `x86`

Author: @bryan-cox Can you confirm whether Kelly's suggestion is accurate?

Reviewer: The recommendation I got from the ARM dev lead was `64-bit ARM` and `64-bit x86` unless it's being typed in a command.
